### ISSUE TITLE
Function requested not supported in SSPI

### ISSUE THREAD

**Issue by lalalandrus** [genotrance/px#3]
I am trying to get this working on Phyton 2.7 64 bit. 64 bit versions of pyutils and pywin32 are installed. The program runs but when I actually get to using the proxy, I get the following error message and a subsequent 407 response from the proxy. 

Traceback (most recent call last):
  File "px.py", line 96, in create_challenge_response
    error_msg, output_buffer = self.sspi_client.authorize(input_buffer)
  File "C:\Python27\lib\site-packages\win32\lib\sspi.py", line 139, in authorize
    sec_buffer_out)
error: (-2146893054, 'InitializeSecurityContext', 'The function requested is not supported')
127.0.0.1 - - [05/Jan/2017 12:02:28] code 407, message Proxy Authentication Required

**Comment by genotrance** [comment]
What version of Windows are you running?

**Comment by lalalandrus** [comment]
Windows 10 64 bit

**Comment by genotrance** [comment]
Are you logging in with an NTLM password (into a domain account since machine is part of the domain) or logging in with a local user account? If it is the latter, there's no SSPI possible from what I understand.

**Comment by lalalandrus** [comment]
It is on a corporate domain.

**Comment by genotrance** [comment]
Can you please confirm the following:-

- Does the binary distribution of Px work on your system? Either 32-bit or 64-bit?
- Does the 32-bit distro of 2.7 Python work on your system?
- How about 3.5 if you get a chance to try that out?

Not really sure why the system isn't working as expected.

**Comment by lalalandrus** [comment]
Thanks for getting back to me. At the moment I have found a workaround that does not require the use of PX. I did however try 32 bit version which fails as well.  It seems to fail the security portion of the proxy. I still kept getting a 407 unauthorized.

**Comment by spequer** [comment]
I got the similar issue here.

Traceback (most recent call last):
  File "D:\tool\python\px-master\px.py", line 100, in create_challenge_response
    error_msg, output_buffer = self.sspi_client.authorize(input_buffer)
  File "C:\Python35\lib\site-packages\win32\lib\sspi.py", line 139, in authorize
    sec_buffer_out)
pywintypes.error: (-2146893048, 'InitializeSecurityContext', 'The token supplied to the function is invalid')

* In a corporation domain
* OS windows7 32bit
* 32bit-distro of 2.7, 3.5

**Comment by genotrance** [comment]
Thanks for the report - this one is slightly different.

Can you please rerun Px with logging enabled in the INI and in a single process (workers=1)? The logs generated might provide some context which will be helpful to figure out what's going on.

**Comment by spequer** [comment]
Oops. I got the answer after google the keyword "sspi SEC_E_INVALID_TOKEN" and following the thread on libcurl: https://curl.haxx.se/mail/lib-2014-09/0044.html

It seems related to "scflags, # security context flags" of sspi.ClientAuth.

 I set scflags to 0 and px works well.

**Comment by genotrance** [comment]
Great catch! I presume you are changing the following in Px (class NtlmMessageGenerator):-

`self.sspi_client = sspi.ClientAuth("NTLM", user)`

to:-

`self.sspi_client = sspi.ClientAuth("NTLM", user, scflags=0)` 

I tried it out and it still works for me, although it works for me even without this. I checked Curl's implementation and they have picked the same in their call to InitializeSecurityContext():-

https://github.com/fl4re/curl/blob/master/lib/curl_sasl_sspi.c

I'll check this in shortly.
